### PR TITLE
Claim account

### PR DESCRIPTION
**Pull request by m4m3m2m1m0**
sorrki

### CODE REVIEW

File: server/src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -1,40 +1,174 @@
...
+  }
+
+  async getMagicLink(
+    magicLinkParam: MagicLinkInput,
+  ): Promise<string | ErrorResponse> {
+    const user = await this.usersService.findOne({
+      email: magicLinkParam.email,
+    });
+
+    if (!user || !user.password) {
```
Comment (queros) [inline server/src/modules/auth/auth.service.ts]: Why are you checking the password?

File: server/src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -1,40 +1,174 @@
...
+      return new ErrorResponse(
+        `User with email: "${magicLinkParam.email}" is not registered.`,
+      );
+    }
+
+    const linkId = uuid();
+
+    user.email = magicLinkParam.email;
```
Comment (queros) [inline server/src/modules/auth/auth.service.ts]: I don't get it... if you found the user by this email, why are you set this email again?

File: server/src/modules/auth/auth.resolver.ts [hunk 3]
```diff
@@ -1,15 +1,76 @@
...
+    @Args('magicLinkId') id: string,
+  ): Promise<Token | ErrorResponse> {
+    return this.authService.loginByMagicLink(id);
+  }
+
+  @Mutation(() => MagicLinkUnion)
+  async sendMagicLink(
+    @Args('magicLinkParam') magicLinkParam: MagicLinkInput,
+  ): Promise<SuccessResponse | ErrorResponse> {
+    const magicLinkResponse = await this.authService.getMagicLink(
```
Comment (queros) [inline server/src/modules/auth/auth.resolver.ts]: Couldn't we send an email inside the authService method (eg sendMagicLink) to keep less logic in the resolver?

File: server/src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -1,40 +1,174 @@
...
 @Injectable()
 export class AuthService {
   constructor(
     @Inject(UsersService) private readonly usersService: UsersService,
   ) {}
 
-  async signIn(signInInput: SignInInput): Promise<TokenModel> {
-    switch (signInInput.grantType) {
+  async registerUser(dto: RegisterUserInput): Promise<Token | ErrorResponse> {
+    const user = await this.usersService.findById(dto._id);
```
Comment (queros) [inline server/src/modules/auth/auth.service.ts]: Where this ID come from? Where do you add a new account?

File: server/src/modules/auth/models/register-user.input.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { Field, InputType } from 'type-graphql';
+import { ObjectIdScalar } from '../../common/graphql-scalars/object-id.scalar';
+import { ObjectId } from 'bson';
+
+@InputType()
+export class RegisterUserInput {
+  @Field(() => ObjectIdScalar)
+  readonly _id: ObjectId;
```
Comment (queros) [inline server/src/modules/auth/models/register-user.input.ts]: I'm not sure about this 🤔

File: server/src/modules/common/common.module.ts [hunk 6]
```diff
@@ -3,5 +3,6 @@ import { ObjectIdScalar } from './graphql-scalars/object-id.scalar';
 
 @Module({
   providers: [ObjectIdScalar],
+  exports: [],
```
Comment (queros) [inline server/src/modules/common/common.module.ts]: For what?

**Comment by Siemko** [comment]
Ok, so we talked it through and came up with an idea to make it simpler and easier for users. The flow will be like that:
1. On the first visit, user creates an account with "username". It is saved as their account as long as they have refresh token in LocalStorage.
1a. On revisit, if refresh token is no longer valid or doesn't exist, user can create an account with "username", same as on the first visit.

2. User can "claim" an account by providing an email address. Account is validated by a link in the first email.
2a. Claimed account is valid, only when the user clicked the link in the first email message.
2b. If the user didn't click the link, they can still use their account as long as they have valid refresh token.

3. When user has validated their account, they can use "send a magic link" button to send them a link to login to their personal account.

We will not use passwords at all, just magic links. We can always add "regular" registration later if needed.

For now we only want magic links.

@m4m3m2m1m0 if you need any advice on that, we can talk it through, also you can ask Kamil L. for help.

**Comment by Siemko** [comment]
PS. Sorrki.

File: server/src/modules/auth/auth.service.ts [hunk 7]
```diff
@@ -1,40 +1,184 @@
...
+    user.activationLinkId = linkId;
+
+    await this.usersService.edit(user);
+
+    const activationLink = `${process.env.BASE_URL}/activateAccount/${linkId}`;
+    const message: SendgridMessage = {
+      from: process.env.FROM_EMAIL_ADDRESS,
+      to: user.email,
+      html: HtmlEmailHelper.generateHtmlEmail(
+        `<h3> Activation link: ${activationLink}</h3>`,
```
Comment (queros) [inline server/src/modules/auth/auth.service.ts]: Change this to a clickable link with `<a>` element, pls.

File: server/src/modules/auth/auth.service.ts [hunk 8]
```diff
@@ -1,40 +1,184 @@
...
+    user.magicLinkId = linkId;
+    user.magicLinkCreatedAt = new Date();
+
+    await this.usersService.edit(user);
+
+    const magicLink = `${process.env.BASE_URL}/link/${linkId}`;
+    const message: SendgridMessage = {
+      from: process.env.FROM_EMAIL_ADDRESS,
+      to: magicLinkParam.email,
+      html: HtmlEmailHelper.generateHtmlEmail(`<h3> Link: ${magicLink}</h3>`),
```
Comment (queros) [inline server/src/modules/auth/auth.service.ts]: Same as above - change this to a clickable link

File: server/package.json [hunk 9]
```diff
@@ -24,14 +24,14 @@
...
-    "argon2": "^0.25.1",
+    "@nestjs/platform-express": "^6.7.2",
+    "@sendgrid/mail": "^6.5.1",
+    "@typegoose/typegoose": "^6.0.4",
+    "apollo-server-express": "^2.9.7",
     "graphql": "^14.5.8",
     "graphql-tools": "^4.0.6",
     "jsonwebtoken": "^8.5.1",
-    "mongoose": "^5.8.1",
+    "mongoose": "^5.7.7",
```
Comment (queros) [inline server/package.json]: Why did you downgrade some packages?

**Pull request merged**